doc/source/whatsnew/v0.24.0.rst

```diff
@@ -798,6 +798,36 @@ Bug in :func:`DataFrame.to_dict` raises ``ValueError`` when used with
 
     df.to_dict(orient='index')
 
+.. _whatsnew_0240.api_breaking.incompatible_index_unions
+
+Incompatible Index Type Unions
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+When performing an :func:`Index.union` operation, between objects of incompatible dtypes,
+the result will be a base :class:`Index` of dtype `object`. This behavior holds true for
+unions between :class:`Index` objects that previously would have been prohibited. The dtype
+of empty :class:`Index` objects will now be evaluated before performing union operations
+rather than simply returning the other :class:`Index` object (:issue:`23525`).
+
```

Comment:
We are making things commutative, maybe mention this. What you have is pretty good.

Reply:
done.